src/Common/src/Internal/Cryptography/BCryptAlgorithmCache.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+
+using Microsoft.Win32.SafeHandles;
+using NTSTATUS = Interop.BCrypt.NTSTATUS;
+using BCryptOpenAlgorithmProviderFlags = Interop.BCrypt.BCryptOpenAlgorithmProviderFlags;
+
+namespace Internal.Cryptography
+{
+    internal static class BCryptAlgorithmCache
+    {
+        /// <summary>
+        ///     Return a SafeBCryptAlgorithmHandle of the desired algorithm and flags. This is a shared handle so do not dispose it!
+        /// </summary>
+        public static SafeBCryptAlgorithmHandle GetCachedBCryptAlgorithmHandle(string hashAlgorithmId, BCryptOpenAlgorithmProviderFlags flags)
+        {
+            // There aren't that many hash algorithms around so rather than use a LowLevelDictionary and guard it with a lock,
+            // we'll use a simple list. To avoid locking, we'll recreate the entire list each time an entry is added and replace it atomically.
+            //
+            // This does mean that on occasion, racing threads may create two handles of the same type, but this is ok.
+
+            // Latch the _cache value into a local so we aren't disrupted by concurrent changes to it.
+            Entry[] cache = _cache;
+            foreach (Entry entry in cache)
+            {
+                if (entry.HashAlgorithmId == hashAlgorithmId && entry.Flags == flags)
+                    return entry.Handle;
+            }
+
+            SafeBCryptAlgorithmHandle safeBCryptAlgorithmHandle;
+            NTSTATUS ntStatus = Interop.BCrypt.BCryptOpenAlgorithmProvider(out safeBCryptAlgorithmHandle, hashAlgorithmId, null, flags);
+            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+
+            Entry[] newCache = new Entry[cache.Length + 1];
+            Entry newEntry = new Entry(hashAlgorithmId, flags, safeBCryptAlgorithmHandle);
+            Array.Copy(cache, newCache, cache.Length);
+
+            // Check again in case some other thread added the hash we need just now. It's fine if he did, but we don't want to create
+            // a duplicate entry in the new cache as that'll slow down all future lookups.
+            foreach (Entry entry in newCache)
```

Comment:
I don't see how this could hit.  You're looking at the same contents as the previous foreach due to the method-local reference and atomic replacement.